### DISCUSSION TITLE
External files for boundary conditions MF6

### DISCUSSION THREAD

**Discussion by laat0003** [modflowpy/flopy#1434] [Q&A] [answered]
Hi,

MF6 is capable of reading stress period data for boundary conditions from external files but I can't seem to find a way to get Flopy to create an appropriate input file. I'm trying to get Flopy to write a GHB input file that accesses an external file via OPEN/CLOSE for stress period data. I've tried using the same approach for external arrays with the NPF package but no love. Am I correct in assuming this functionality is yet to be implemented?

Cheers.

**Comment by christianlangevin** [comment, marked as answer]
Hi @laat0003, all this should be implemented and fully functional for MODFLOW 6, though it seems our documentation is a bit scattered.  There are some details described in this [online doc](https://flopy.readthedocs.io/en/3.3.5/tutorials.html#modflow-6-tutorials) and in this [notebook](https://github.com/modflowpy/flopy/blob/develop/examples/Notebooks/flopy3_mf6_tutorial.ipynb).

Here is a running example, based on the example on main FloPy page, that shows how to write data to external files.

```
import flopy
ws = './mymodel'
name = 'mymodel'
sim = flopy.mf6.MFSimulation(sim_name=name, sim_ws=ws, exe_name='mf6')
tdis = flopy.mf6.ModflowTdis(sim)
ims = flopy.mf6.ModflowIms(sim)
gwf = flopy.mf6.ModflowGwf(sim, modelname=name, save_flows=True)
dis = flopy.mf6.ModflowGwfdis(gwf, nrow=10, ncol=10)
ic = flopy.mf6.ModflowGwfic(gwf)

# There is full control over the grid data arrays, including
# saving the arrays in binary form to external files
k11_values = np.full((1, 10, 10), 10.0)
k11 = {
    "filename": "k11.txt",
    "factor": 1.0,
    "data": k11_values,
    "iprn": 1,
    "binary": "True",
}
npf = flopy.mf6.ModflowGwfnpf(gwf, k=k11, save_specific_discharge=True)

# The transient list data can also be stored in external files by
# calling the store_as_external_file() method on the stress period
# data.  The list data can also be optionally stored in binary format
# in some cases.
chd = flopy.mf6.ModflowGwfchd(gwf, stress_period_data=[[(0, 0, 0), 1.],
                                                       [(0, 9, 9), 0.]])
chd.stress_period_data.store_as_external_file('mychds.dat', binary=True)

budget_file = name + '.bud'
head_file = name + '.hds'
oc = flopy.mf6.ModflowGwfoc(gwf,
                            budget_filerecord=budget_file,
                            head_filerecord=head_file,
                            saverecord=[('HEAD', 'ALL'), ('BUDGET', 'ALL')])
sim.write_simulation()
sim.run_simulation()
```

**Reply by laat0003** [reply]
Brilliant! That's exactly what I was after. I should have checked the other notebooks. I was looking at the notebook on external file handling and didn't see it in there, so assumed it wasn't implemented yet. You know what they say about making assumptions,... they make an ass out of me and... no wait,... how does that go again?
